File: Java/9.시뮬레이션/아기상어16236.cpp
/*
#include <vector>
#include <algorithm>
#include <iostream>
#include <queue>
#include <string.h>
#include <tuple>
using namespace std;
int dx[] = {-1,0,1,0};
int dy[] = {0,1,0,-1};
int dist[21][21];
int move(vector<vector<int>> &a,int &x,int &y,int size)
{
    int n = a.size();
    memset(dist,-1,sizeof(dist));
    queue<pair<int,int>> q;
    q.push(make_pair(x,y));
    dist[x][y] = 0;
    while(!q.empty())
    {
        int x,y;
        tie(x,y) = q.front();
        q.pop();
        for(int i=0;i<4;i++)
        {
            int nx = x+dx[i];
            int ny = y+dy[i];
            if(nx>=0&&nx<n&&ny>=0&&ny<n&&dist[nx][ny]==-1)
            {
                if(a[nx][ny]<=size)
                {
                    q.push(make_pair(nx,ny));
                    dist[nx][ny] = dist[x][y]+1;
                }
            }
        }
    }
    int mini = 99999999;
    int ex,ey;
    for(int i=0;i<n;i++)
    {
        for(int j=0;j<n;j++)
        {
            if(a[i][j]!=0&&a[i][j]!=9&&a[i][j]<size&&dist[i][j]!=-1)
            {
                if(dist[i][j] < mini)
                {
                    mini = dist[i][j];
                    ex = i;
                    ey = j;
                }
            }
        }
    }
    if(mini==99999999) return -1;
    a[ex][ey] = 0;
    x = ex; y = ey;
    return dist[ex][ey];
}
int main()
{
    int n;
    cin >> n;
    int sx,sy;
    vector<vector<int>> a(n,vector<int>(n,0));
    for(int i=0;i<n;i++)
    {
        for(int j=0;j<n;j++)
        {
            cin >> a[i][j];
            if(a[i][j]==9)
            {
                a[i][j] = 0;
                sx = i;
                sy = j;
            }
        }
    }
    int size = 2;
    int cnt = 0;
    int eat = 0;
    while(true)
    {       
        int time = move(a,sx,sy,size);
        if(time==-1) break;
        cnt += time;
        eat += 1;
        if(eat==size)
        {
            size++;
            eat = 0;
        }
    }
    cout << cnt << '\n';
    return 0;
}
*/
#include <vector>
#include <algorithm>
#include <iostream>
#include <queue>
#include <string.h>
#include <tuple>
using namespace std;
int dx[] = {0,-1,0,1};
int dy[] = {-1,0,1,0};
int a[21][21];
int dist[21][21];
int N;
int move(int &x,int &y,int level)
{
    queue<pair<int,int>> q;
    q.push(make_pair(x,y));
    dist[x][y] = 0;
    while(!q.empty())
    {
        int x,y;
        tie(x,y) = q.front();
        q.pop();
        for(int i=0;i<4;i++)
        {
            int nx = x+dx[i];
            int ny = y+dy[i];
            if(nx>=0&&nx<N&&ny>=0&&ny<N&&dist[nx][ny]==-1)
            {
                if(a[nx][ny]<=level)
                {
                    q.push(make_pair(nx,ny));
                    dist[nx][ny] = dist[x][y] + 1;
                }
            }
        }
    }
    int mini = -1;
    int fx,fy;
    for(int i=0;i<N;i++)
    {
        for(int j=0;j<N;j++)
        {
            if(mini==-1||mini>dist[i][j])
            {
                if(dist[i][j]!=-1&&a[i][j]!=0&&a[i][j]<level)
                {
                    mini = dist[i][j];
                    fx = i;
                    fy = j;
                }
            }
        }
    }
    if(mini==-1) return -1;
    else
    {
        a[fx][fy] = 0;
        x = fx;
        y = fy;
        return mini;
    }
}
int main()
{
    int n;
    cin >> n;
    N=n;
    int sx,sy;
    for(int i=0;i<n;i++)
    {
        for(int j=0;j<n;j++)
        {
            cin >> a[i][j];
            if(a[i][j]==9)
            {
                sx = i;
                sy = j;
                a[i][j] = 0;
            }
        }
    }
    int answer = 0;
    int level = 2;
    int cnt = 0;
    while(true)
    {
        memset(dist,-1,sizeof(dist));
        int k = move(sx,sy,level);
        if(k==-1) break;
        answer += k;
        cnt++;
        if(cnt==level)
        {
            level++;
            cnt = 0;
        }
    }
    cout << answer << '\n';
    return 0;
}















/*
    상어가 최단거리의 물고기를 먹을수 있는 경우를 구하는데 예외처리를 못해 오래걸림.
*/

File: Java/9.시뮬레이션/아기상어16236.java
//import java.io.BufferedReader;
//import java.io.IOException;
//import java.io.InputStreamReader;
//import java.util.ArrayList;
//import java.util.Arrays;
//import java.util.Collections;
//import java.util.HashSet;
//import java.util.LinkedList;
//import java.util.List;
//import java.util.Queue;
//import java.util.StringTokenizer;
//
//public class Main {
//
//	static int[] dx = {0,-1,0,1};
//	static int[] dy = {-1,0,1,0};
//	static int[][] board = new int[21][21];
//	static int[][] dist = new int[21][21];
//	static int sx,sy;
//	static int N;
//
//	static int move(int level) {
//		for(int i=0;i<dist.length;i++) Arrays.fill(dist[i], Integer.MAX_VALUE);
//		Queue<int[]> q = new LinkedList<>();
//		q.add(new int[] {sx,sy});
//		dist[sx][sy] = 0;
//		while(!q.isEmpty()) {
//			int[] now = q.poll();
//			int x = now[0]; int y = now[1];
//			for(int i=0;i<4;i++) {
//				int nx = x+dx[i];
//				int ny = y+dy[i];
//				if(nx>=0&&nx<N&&ny>=0&&ny<N&&dist[nx][ny]==Integer.MAX_VALUE) {
//					if(board[nx][ny]<=level) {
//						q.add(new int[] {nx,ny});
//						dist[nx][ny] = dist[x][y]+1;
//					}
//				}
//			}
//		}
//		int mini = Integer.MAX_VALUE;
//		int mx = -1; int my = -1;
//		for(int i=0;i<N;i++) {
//			for(int j=0;j<N;j++) {
//				if(dist[i][j]<mini&&board[i][j]<level&&board[i][j]!=0) {
//					mini = dist[i][j];
//					mx = i;
//					my = j;
//				}
//			}
//		}
//		if(mx==-1&&my==-1) return 0;
//		else {
//			board[mx][my] = 0;
//			sx = mx;
//			sy = my;
//			return mini;
//		}
//
//	}
//
//	public static void main(String[] args) throws IOException {
//		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//		StringBuilder sb = new StringBuilder();
//		StringTokenizer st;
//
//		int n = Integer.parseInt(br.readLine());
//		N=n;
//		for(int i=0;i<n;i++) {
//			st = new StringTokenizer(br.readLine());
//			for(int j=0;j<n;j++) {
//				board[i][j] = Integer.parseInt(st.nextToken());
//				if(board[i][j]==9) {
//					sx = i;
//					sy = j;
//					board[i][j] = 0;
//				}
//			}
//		}
//		int answer = 0;
//		int level = 2;
//		int cnt = 0;
//		while(true) {
//			int time = move(level);
//			if(time==0) {
//				break;
//			}
//			answer+=time;
//			cnt+=1;
//			if(cnt==level) {
//				level+=1;
//				cnt=0;
//			}
//		}
//		System.out.println(sb.append(answer));
//	}
//}
import java.io.*;
import java.util.*;

public class Main {

	static int[] dx = {0,-1,0,1};
	static int[] dy = {-1,0,1,0};
	static int[][] board = new int[21][21];
	static int[][] dist = new int[21][21];
	static int N;
	static int sx=-1,sy=-1;
	
	static int move(int level) {
		for(int i=0;i<N;i++) {
			for(int j=0;j<N;j++) dist[i][j] = 0;
		}
		Queue<int[]> q = new LinkedList<>();
		q.offer(new int[] {sx,sy});
		dist[sx][sy] = 1;
		while(!q.isEmpty()) {
			int[] now = q.poll();
			int x = now[0]; int y = now[1];
			for(int i=0;i<4;i++) {
				int nx = x+dx[i];
				int ny = y+dy[i];
				if(nx>=0&&nx<N&&ny>=0&&ny<N&&dist[nx][ny]==0) {
					if(board[nx][ny]<=level) {
						q.offer(new int[] {nx,ny});
						dist[nx][ny] = dist[x][y]+1;
					}
				}
			}
		}
		int mini = Integer.MAX_VALUE;
		int min_x = -1, min_y = -1;
		for(int i=0;i<N;i++) {
			for(int j=0;j<N;j++) {
				if(board[i][j]!=0&&board[i][j]<level&&dist[i][j]!=0) {
					if(mini>dist[i][j]) {
						mini = dist[i][j];
						min_x=i;min_y=j;
					}
				}
			}
		}
		if(mini==Integer.MAX_VALUE) return -1;
		else {
			board[min_x][min_y] = 0;
			sx = min_x; sy = min_y;
			return mini-1;
		}
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;
		
		int n = Integer.parseInt(br.readLine());
		N=n;
		for(int i=0;i<n;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<n;j++) {
				board[i][j] = Integer.parseInt(st.nextToken());
				if(board[i][j]==9) {
					sx=i;sy=j;
					board[i][j] = 0;
				}
			}
		}
		int level = 2;
		int cnt = 0;
		int time = 0;
		while(true) {
			int result = move(level);
			if(result==-1) break;
			time+=result;
			cnt+=1;
			if(cnt==level) {
				level += 1;
				cnt = 0;
			}
		}
		System.out.println(sb.append(time));
	}
}
